### PR TITLE
In memory parallel read.

### PR DESCRIPTION
**Pull request by roigcarlo**
I cleaned it up a bit since basically was already working.
Also I made the process derive from `metis_divide_heterogeneous_input_process` instead of `metis_divide_input_process` so most of the code is reused.

### CODE REVIEW

File: kratos/sources/model_part_io.cpp [hunk 1]
```diff
@@ -70,12 +70,12 @@ namespace Kratos
       , mOptions(Options)
     {
         // nullptr test can be confusing with Kratos::shared_ptr. Commented until we move to std::shared_ptr
```
Comment (jcotela) [inline kratos/sources/model_part_io.cpp]: this comment is now obsolete...

File: applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h [hunk 2]
```diff
@@ -166,186 +162,187 @@ class MetisDivideHeterogeneousInputInMemoryProcess : public MetisDivideInputToPa
...
+            IO::ConnectivitiesContainerType KratosFormatNodeConnectivities;
 
-        SizeType NumNodes = BaseType::mrIO.ReadNodalGraph(KratosFormatNodeConnectivities);
+            SizeType NumNodes = BaseType::mrIO.ReadNodalGraph(KratosFormatNodeConnectivities);
 
-          // Write connectivity data in CSR format
-        idxtype* NodeIndices = 0;
-        idxtype* NodeConnectivities = 0;
+            // Write connectivity data in CSR format
+            idxtype* NodeIndices = 0;
```
Comment (RiccardoRossi) [inline applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h:174]: why don't you use a std::vector here? (you can still pass the data() field if you need to pass it as c array)

File: applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h [hunk 3]
```diff
@@ -166,186 +162,187 @@ class MetisDivideHeterogeneousInputInMemoryProcess : public MetisDivideInputToPa
...
-      }
+        if(mpi_rank != 0) {
+            streams[mpi_rank]->write(mpi_recv_buffer[mpi_rank], msgRecvSize[mpi_rank]);
+        }
 
-      std::ofstream* p_ofstream = new std::ofstream("debug_modelpart_"+std::to_string(mpi_rank));
-      (*p_ofstream) << stringbufs[mpi_rank].str() << std::endl;
+#ifdef KRATOS_DEBUG
+        // Print the partitions in debug mode
+        std::ofstream* p_ofstream = new std::ofstream("debug_modelpart_"+std::to_string(mpi_rank)+".mpda");
```
Comment (msandre) [inline applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h]: Memory leak of p_ofstream.

Comment (roigcarlo) [reply applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h]: Thx, Fixed!

File: kratos/sources/model_part_io.cpp [hunk 4]
```diff
@@ -69,13 +69,13 @@ namespace Kratos
...
-        // if (Stream == nullptr)
-        //    KRATOS_THROW_ERROR(std::invalid_argument, "Error: ModelPartIO Stream is invalid ", "");
+        // Check if the pointer is valid
+        if (Stream == nullptr)
+           KRATOS_THROW_ERROR(std::invalid_argument, "Error: ModelPartIO Stream is invalid ", "");
 
         // Check if the pointer was .reset() or never initialized and if its a NULL pointer)
-        // if (Stream == NULL || Stream == Kratos::shared_ptr<std::iostream>(NULL))
-        //    KRATOS_THROW_ERROR(std::invalid_argument, "Error: ModelPartIO Stream is invalid ", "");
+        if (Stream == NULL || Stream == Kratos::shared_ptr<std::iostream>(NULL))
```
Comment (RiccardoRossi) [inline kratos/sources/model_part_io.cpp]: i would use nullptr here too

File: applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h [hunk 5]
```diff
@@ -166,186 +162,187 @@ class MetisDivideHeterogeneousInputInMemoryProcess : public MetisDivideInputToPa
...
-            Msg << std::endl;
-            Msg << "ERROR in MetisDivideHeterogenousInputProcess:" << std::endl;
-            Msg << "Read " << NumElements << " elements, but element list has " << ElementConnectivities.size() << " entries." << std::endl;
-            Msg << "Elements are most likely not correlatively numbered." << std::endl;
+            // Partition elements
+            IO::ConnectivitiesContainerType ElementConnectivities;
+            SizeType NumElements =  BaseType::mrIO.ReadElementsConnectivities(ElementConnectivities);
+            if (NumElements != ElementConnectivities.size())
+            {
+                std::stringstream Msg;
```
Comment (RiccardoRossi) [inline applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h:191]: very minor...but here you could use directly kratos_error

File: applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h [hunk 6]
```diff
@@ -166,186 +162,187 @@ class MetisDivideHeterogeneousInputInMemoryProcess : public MetisDivideInputToPa
...
+            int NumColors;
+            GraphType ColoredDomainGraph;
+            GraphColoringProcess(mNumberOfPartitions,DomainGraph,ColoredDomainGraph,NumColors).Execute();
 
-        // Create lists containing all nodes/elements/conditions known to each partition
-        DividingNodes(nodes_all_partitions, ElementConnectivities, ConditionConnectivities, NodePartition, ElementPartition, ConditionPartition);
-        DividingElements(elements_all_partitions, ElementPartition);
-        DividingConditions(conditions_all_partitions, ConditionPartition);
+            if (mVerbosity > 0) {
+                KRATOS_WATCH(NumColors);
```
Comment (RiccardoRossi) [inline applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h]: info

File: applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h [hunk 7]
```diff
@@ -166,186 +162,187 @@ class MetisDivideHeterogeneousInputInMemoryProcess : public MetisDivideInputToPa
...
-        if (mVerbosity > 1)
-        {
-            std::cout << "Final list of nodes known by each partition" << std::endl;
-            for(SizeType i = 0 ; i < NumNodes ; i++)
-            {
-                std::cout << "Node #" << i+1 << "->";
-                for(std::vector<std::size_t>::iterator j = nodes_all_partitions[i].begin() ; j != nodes_all_partitions[i].end() ; j++)
-                    std::cout << *j << ",";
-                std::cout << std::endl;
+            if (mVerbosity > 2) {
```
Comment (RiccardoRossi) [inline applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h:245]: info

File: applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h [hunk 8]
```diff
@@ -166,186 +162,187 @@ class MetisDivideHeterogeneousInputInMemoryProcess : public MetisDivideInputToPa
...
+#endif
 
-      // TODO: Try to come up with a better way to change the buffer.
-      ((ModelPartIO&)mrIO).SwapStreamSource(streams[mpi_rank]);
+        // TODO: Try to come up with a better way to change the buffer.
+        ((ModelPartIO&)mrIO).SwapStreamSource(streams[mpi_rank]);
 
-      // Free buffers
-      free(mpi_recv_buffer[mpi_rank]);
+        // Free buffers
```
Comment (RiccardoRossi) [inline applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h:334]: do this with a std vector (use the .data() in the mpi calls

File: applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h [hunk 9]
```diff
@@ -166,186 +162,187 @@ class MetisDivideHeterogeneousInputInMemoryProcess : public MetisDivideInputToPa
...
 
-      delete [] reqs;
-      delete [] stats;
+        delete [] reqs;
+        delete [] stats;
 
-      delete [] mpi_recv_buffer;
-      delete [] mpi_send_buffer;
-      delete [] str;
+        delete [] mpi_recv_buffer;
```
Comment (RiccardoRossi) [inline applications/metis_application/custom_processes/metis_divide_heterogeneous_input_in_memory_process.h:340]: i woukd suggest using std vector whenever possible

**Comment by roigcarlo** [comment]
@RiccardoRossi I will deal with the logging issues now. About the usage of std::vector... I also preffer to use stl but I did it this way mainly because the rest of metis application ( and kratos) works this way.

Imho if we want to use stl instead of C arrays we should start changing it in a separate PR for all the processes in the metis application, but if it is really necessary I'll do it here.

**Comment by RiccardoRossi** [comment]
@roigcarlo i am not blocking this because of std::vector, it was simply a suggestion

**Comment by RiccardoRossi** [comment]
did u push the logger changes?

**Comment by roigcarlo** [comment]
Seems I forgot to upload them @RiccardoRossi sorry :(

**Review by RiccardoRossi — APPROVED** [review]

**Pull request merged**